presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -128,51 +110,42 @@ private ConnectorSplit toIcebergSplit(TupleDomain<HiveColumnHandle> predicate, F
                 task.start(),
                 task.length(),
                 ImmutableList.of(),
-                nameToId,
                 predicate,
                 getPartitionKeys(task));
     }
 
-    private static List<HivePartitionKey> getPartitionKeys(FileScanTask scanTask)
+    private static List<IcebergPartitionKey> getPartitionKeys(FileScanTask scanTask)
     {
         StructLike partition = scanTask.file().partition();
         PartitionSpec spec = scanTask.spec();
         Map<PartitionField, Integer> fieldToIndex = getIdentityPartitions(spec);
-        List<HivePartitionKey> partitionKeys = new ArrayList<>();
+        List<IcebergPartitionKey> partitionKeys = new ArrayList<>();
 
         fieldToIndex.forEach((field, index) -> {
-            String name = field.name();
-            Type sourceType = spec.schema().findType(field.sourceId());
-            Type partitionType = field.transform().getResultType(sourceType);
-            Class<?> javaClass = partitionType.typeId().javaClass();
+            int id = field.sourceId();
+            Type type = spec.schema().findType(id);
+            Class<?> javaClass = type.typeId().javaClass();
             Object value = partition.get(index, javaClass);
-            String partitionValue = HIVE_DEFAULT_DYNAMIC_PARTITION;
-            if (value != null) {
-                switch (partitionType.typeId()) {
-                    case DATE:
-                        partitionValue = DATE_FORMATTER.format(LocalDate.ofEpochDay((int) value));
-                        break;
-                    case TIMESTAMP:
-                        partitionValue = TIMESTAMP_FORMATTER.format(toLocalDateTime((long) value));
-                        break;
-                    case FIXED:
-                    case BINARY:
-                        partitionValue = new String(((ByteBuffer) value).array(), UTF_8);
-                        break;
-                    default:
-                        partitionValue = value.toString();
-                }
+
+            if (value == null) {
+                throw new PrestoException(ICEBERG_INVALID_PARTITION_VALUE, format(
+                        "File %s has no partition data for partitioning column %s",
+                        scanTask.file().path().toString(),
+                        field.name()));
+            }
+
+            String partitionValue;
+            switch (type.typeId()) {
+                case FIXED:
+                case BINARY:
+                    partitionValue = new String(((ByteBuffer) value).array(), UTF_8);
+                    break;
+                default:
+                    partitionValue = value.toString();
             }
-            partitionKeys.add(new HivePartitionKey(name, partitionValue));
```

Comment:
how about storing a deserialized presto type and value object here itself? it reduces the propagation of untyped values a bit

Reply:
This is a good idea but for now the value stored in `IcebergPartitionKey` is not technically a "presto" value. It's more an "Iceberg" value. The deserializer (`IcebergPageSource#deserializePartitionValue`) not only deserializes but also converts values from Iceberg representation to Presto representation.

I think we should do the conversion firstly and then serialize/deserialize Presto values. But I prefer to do it in a future separate PR. Because the value conversion also relates to `DomainConverter` and `ExpressionConverter`. We should do a uniform refactoring in the future.